### PR TITLE
CLI Show Workflow Namespaces Option

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

- [X] Adds the ability to get the objects and their namespaces from a workflow and present them as a table
- [X] Adds the option flag `--get-namespaces` to execute this behavior from the command line with an Arcaflow executable

## Behavior Happy Path

```shell
❯ ./arcaflow --get-namespaces --workflow workflow.yaml --config config.yaml --input input.yaml
OBJECT              NAMESPACE   
HogCpuInput          $.steps.hog_cpu_wf.execute.inputs.items
KubernetesTarget     $.steps.hog_cpu_wf.execute.inputs.items
ThirdObject          $.steps.hog_cpu_wf.execute.inputs.items
StressNGParams       $.steps.hog_cpu_wf.execute.inputs.items.stressng_params
error                $.steps.hog_cpu_wf.failed.outputs.error
data                 $.steps.hog_cpu_wf.outputs.outputs.success
```

given `workflow.yaml`
```yaml
version: v0.2.0
input:
  root: HogCpuInput
  objects:
    ThirdObject:
      id: ThirdObject
      properties:
        name:
          type:
            type_id: string
    KubernetesTarget:
      id: KubernetesTarget
      properties:
        kubeconfig_path:
          type:
            type_id: string
        namespace:
          type:
            type_id: string
          required: true
    HogCpuInput:
      id: HogCpuInput
      properties:
        stressng_params:
          type:
            type_id: ref
            id: StressNGParams
            namespace: $.steps.stressng.starting.inputs.input

steps:
  stressng:
    plugin: 
      src: quay.io/arcalot/arcaflow-plugin-stressng:0.6.0
      deployment_type: image
    step: workload
    input: !expr $.input.stressng_params
            
outputs:
  success:
    stressng: !expr $.steps.stressng.outputs.success

```
---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: cmd/arcaflow/main.go [hunk 1]
```diff
@@ -58,19 +58,22 @@ func main() {
...
-		configUsage   = "The path to the Arcaflow configuration file to load, if any."
-		inputUsage    = "The path to the workflow input file to load, if any."
-		contextUsage  = "The path to the workflow directory to run from."
-		workflowUsage = "The path to the workflow file to load."
+		versionUsage       = "Print Arcaflow Engine version and exit."
+		configUsage        = "The path to the Arcaflow configuration file to load, if any."
+		inputUsage         = "The path to the workflow input file to load, if any."
+		contextUsage       = "The path to the workflow directory to run from."
+		workflowUsage      = "The path to the workflow file to load."
+		getNamespacesUsage = "Show the namespaces available to this workflow."
```
Comment (dbutenhof) [inline cmd/arcaflow/main.go:69]: This thing of keeping the `=` aligned generates a lot of unnecessary whitespace changes. Yeah, we can disable whitespace diffs, but *sometimes* they're relevant and I don't like to start that way. Goes the Go linter really (stupidly 🤣) insist on this?

File: cmd/arcaflow/main.go [hunk 2]
```diff
@@ -197,31 +200,36 @@ func runWorkflow(flow engine.WorkflowEngine, fileCtx loadfile.FileCache, workflo
...
-	)
-	if err != nil {
-		logger.Errorf("Failed to marshal output (%v)", err)
-		return ExitCodeInvalidData
-	}
-	_, _ = os.Stdout.Write(data)
-	if outputError {
-		return ExitCodeWorkflowErrorOutput
+	if getNamespaces {
+		workflow.PrintNamespaceResponse(os.Stdout, workFlowObj.Namespaces(), logger)
```
Comment (dbutenhof) [inline cmd/arcaflow/main.go:210]: Maybe `PrintNamespaceReport` (or "Table") would be better than "Response"? Or even `ReportNamespaces`?

Also, if you added a `return` here, you could avoid the `else` and changing the indent of the lines below. The function already has multiple returns, so that would help rather than hurt readability.

File: internal/tableprinter/tableprinter.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	tabwriterMinWidth = 6
+	tabwriterWidth    = 4
+	tabwriterPadding  = 3
+	tabwriterPadChar  = ' '
+	tabwriterFlags    = tabwriter.FilterHTML
+)
+
+// NewTabWriter returns a tabwriter that transforms tabbed columns into aligned
+// text.
+func NewTabWriter(output io.Writer) *tabwriter.Writer {
```
Comment (dbutenhof) [inline internal/tableprinter/tableprinter.go:22]: Maybe make this `NewTablePrinter` to match the module name, so it's easier to find? Also, this name ties it more closely than necessary to the underlying `text/tabwriter` package, I think.

Then again, should this be capitalized, or is it really just a private helper for `PrintTwoColumnTable`?

File: internal/tableprinter/tableprinter.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+	// column headers are at the top, so they are written first
+	for _, col := range headers {
+		_, _ = fmt.Fprint(w, strings.ToUpper(col), "\t")
+	}
+	_, _ = fmt.Fprintln(w)
+
+	// rows form the body of the table
+	for _, row := range rows {
+		_, _ = fmt.Fprintln(w, row[0], "\t", row[1])
```
Comment (dbutenhof) [inline internal/tableprinter/tableprinter.go:39]: I suppose I could look this up, but ... how smart is this "tab writer" about column alignment? Does the tab separator ensure the columns are aligned, or does this risk misaligning the second column when the first pushes past a tab stop?

Then again, your test output below shows all space characters, which is more or less suggested by your default parameters, so maybe I'm misinterpreting the context of the `"\t"` here?

File: internal/tidy/tidy_test.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	expOut := [][]string{
+		{"astromech", "bb"},
+		{"astromech", "q7"},
+		{"astromech", "r2"},
+		{"astromech", "r4"},
+		{"battle", "b1"},
+		{"battle", "ig"},
+		{"probe", "viper"},
+		{"protocol", "000"},
+		{"protocol", "c3po"},
```
Comment (dbutenhof) [inline internal/tidy/tidy_test.go:36]: I wonder if there's any remote chance of getting in trouble for naming C3PO in your code? You used nicknames for the astromechs, so maybe "3po" would be both safer and more consistent here in any case? 😆

File: internal/tidy/tidy_test.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+func TestSwapColumnsHappy(t *testing.T) {
+	testData := map[string]struct {
+		input    [][]string
+		expected [][]string
+	}{
+		"happy": {
+			input: [][]string{{"fruit", "tomato"},
+				{"fruit", "cucumber"},
+				{"veggie", "spinach"},
+				{"veggie", "carrot"},
```
Comment (dbutenhof) [inline internal/tidy/tidy_test.go:73]: There's no such thing as a "vegetable" ... spinach is a leaf and carrot is a root. Yeah, OK, too much ...

File: internal/tidy/tidy.go
Comment (dbutenhof) [inline internal/tidy/tidy.go:1]: I have no idea why someone named this "tidy", and although I get why you're carrying that for your derivative, I don't like it much. Ah, well ...

File: cmd/arcaflow/main.go (hunk 1 above)
Comment (jaredoconnell) [reply cmd/arcaflow/main.go:69]: It's definitely nicer to look at, with the only downside being the diff's readability.

**Review by jaredoconnell — APPROVED** [review]
I think this will be ready once Dave's concerns are addressed.

File: internal/tidy/tidy_test.go (hunk 5 above)
Comment (mfleader) [reply internal/tidy/tidy_test.go:36]: yeah I had a feeling the tm might be an issue; I'll change that

File: internal/tidy/tidy.go
Comment (mfleader) [reply internal/tidy/tidy.go:1]: The domain specific language is not specific to R. It's just more common in the R community. It's from the [2014 Tiday Data paper](https://vita.had.co.nz/papers/tidy-data.pdf).

File: cmd/arcaflow/main.go (hunk 1 above)
Comment (mfleader) [reply cmd/arcaflow/main.go:69]: Yeah I just accepted this formatting from goland when I started.

File: internal/tableprinter/tableprinter.go (hunk 4 above)
Comment (mfleader) [reply internal/tableprinter/tableprinter.go:39]: I think it does risk pushing past the tab alignment. From my example at the top, the tab separator does not ensure they are perfectly aligned, but they will be nearly aligned which seems good enough.

```shell
OBJECT              NAMESPACE   
HogCpuInput          $.steps.hog_cpu_wf.execute.inputs.items
```

File: internal/tableprinter/tableprinter.go (hunk 3 above)
Comment (mfleader) [reply internal/tableprinter/tableprinter.go:22]: I didn't use `NewTablePrinter` because the function returns a `tabwriter.Writer`, and I did not think it was worth the effort to create a new class called `TablePrinter`.

At the moment, yes the tabwriter only exists to print this table. I can make it private.

File: cmd/arcaflow/main.go (hunk 2 above)
Comment (mfleader) [reply cmd/arcaflow/main.go:210]: I changed it to `PrintObjectNamespaceTable`.

Comment (mfleader) [reply cmd/arcaflow/main.go]: Added the `return`

File: internal/tidy/tidy.go
Comment (mfleader) [reply internal/tidy/tidy.go:1]: I think that if I'm going to use their dsl to describe the behavior in my package then it makes sense to use a similar package name as well. I disagree with the reasoning for changing the package name, but I think `tablefmt` is a good enough name, and I also don't think it's worth investing time in a debate over its name.

**Review by dbutenhof — APPROVED** [review]
